pkgs/test/nixpkgs-check-by-name/src/nix_file.rs

```diff
@@ -0,0 +1,508 @@
+    /// attempt at distinguishing this.
+    pub fn call_package_argument_info_at(
+        &self,
+        line: usize,
+        column: usize,
+        relative_to: &Path,
+    ) -> anyhow::Result<Option<CallPackageArgumentInfo>> {
+        let Some(attrpath_value) = self.attrpath_value_at(line, column)? else {
+            return Ok(None);
+        };
+        self.attrpath_value_call_package_argument_info(attrpath_value, relative_to)
+    }
+
+    // Internal function mainly to make it independently testable
+    fn attrpath_value_at(
+        &self,
+        line: usize,
+        column: usize,
+    ) -> anyhow::Result<Option<ast::AttrpathValue>> {
+        let index = self.line_index.fromlinecolumn(line, column);
+
+        let token_at_offset = self
+            .syntax_root
+            .syntax()
+            .token_at_offset(TextSize::from(index as u32));
+
+        // The token_at_offset function takes indices to mean a location _between_ characters,
+        // which in this case is some spacing followed by the attribute name:
+        //
+        //   foo = 10;
+        //  /\
+        //  This is the token offset, we get both the (newline + indentation) on the left side,
+        //  and the attribute name on the right side.
+        let TokenAtOffset::Between(_space, token) = token_at_offset else {
+            anyhow::bail!("Line {line} column {column} in {} is not the start of a token, but rather {token_at_offset:?}", self.path.display())
+        };
+
+        // token looks like "foo"
+        let Some(node) = token.parent() else {
+            anyhow::bail!(
+                "Token on line {line} column {column} in {} does not have a parent node: {token:?}",
+                self.path.display()
+            )
+        };
+
+        // node looks like "foo"
+        let Some(attrpath_node) = node.parent() else {
+            anyhow::bail!(
+                "Node in {} does not have a parent node: {node:?}",
+                self.path.display()
+            )
+        };
+
+        if attrpath_node.kind() != SyntaxKind::NODE_ATTRPATH {
+            return Ok(None);
+        }
+        // attrpath_node looks like "foo.bar"
+        let Some(attrpath_value_node) = attrpath_node.parent() else {
+            anyhow::bail!(
+                "Attribute path node in {} does not have a parent node: {attrpath_node:?}",
+                self.path.display()
+            )
+        };
+
+        if !ast::AttrpathValue::can_cast(attrpath_value_node.kind()) {
+            anyhow::bail!(
+                "Node in {} is not an attribute path value node: {attrpath_value_node:?}",
+                self.path.display()
+            )
+        }
+        // attrpath_value_node looks like "foo.bar = 10;"
+
+        // unwrap is fine because we confirmed that we can cast with the above check
+        Ok(Some(ast::AttrpathValue::cast(attrpath_value_node).unwrap()))
```

Comment:
Unfortunately `cast` consumes `attrpath_value_node`, so if we try to do

```rust
        if let Some(attrpath_value_ast_node) = ast::AttrpathValue::cast(attrpath_value_node) {
            Ok(Some(attrpath_value_ast_node))
        } else {
            anyhow::bail!(
                "Node in {} is not an attribute path value node: {attrpath_value_node:?}",
                self.path.display()
            )
        }
```

We'll get a borrow checker error:

```
    Checking nixpkgs-check-by-name v0.1.0 (/home/tweagysil/src/nixpkgs/by-name/pkgs/test/nixpkgs-check-by-name)
error[E0382]: borrow of moved value: `attrpath_value_node`
   --> src/nix_file.rs:180:66
    |
169 |         let Some(attrpath_value_node) = attrpath_node.parent() else {
    |                  ------------------- move occurs because `attrpath_value_node` has type `rowan::SyntaxNode<NixLanguage>`, which does not implement the `Copy` trait
...
176 |         if let Some(attrpath_value_ast_node) = ast::AttrpathValue::cast(attrpath_value_node) {
    |                                                                         ------------------- value moved here
...
180 |                 "Node in {} is not an attribute path value node: {attrpath_value_node:?}",
    |                                                                  ^^^^^^^^^^^^^^^^^^^^^^^ value borrowed here after move
    |
    = note: this error originates in the macro `$crate::__export::format_args` which comes from the expansion of the macro `anyhow::bail` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider cloning the value if the performance cost is acceptable
    |
176 |         if let Some(attrpath_value_ast_node) = ast::AttrpathValue::cast(attrpath_value_node.clone()) {
    |                                                                                            ++++++++
```


Could do the `clone` as suggested by the error message, but feels a bit pointless to clone when it shouldn't be necessary. I believe the `rnix` crate should really have a way to work with non-owned values.

Reply:
For now I'm adding a comment explaining this here.